okio/src/commonMain/kotlin/okio/internal/-Path.kt

```diff
@@ -159,6 +155,84 @@ internal inline fun Path.commonResolve(child: Path): Path {
   return buffer.toPath()
 }
 
+@ExperimentalFileSystem
+@Suppress("NOTHING_TO_INLINE")
+internal inline fun Path.commonRelativeTo(other: Path): Path {
+  require(isAbsolute && other.isAbsolute || isRelative && other.isRelative) {
+    "Paths of different roots cannot be relative to each other: $this and $other"
+  }
+
+  val thisSlash = slash
+  val otherSlash = other.slash
+  require(thisSlash == otherSlash) {
+    "Paths of different platforms cannot be relative to each other: $this and $other"
+  }
+
+  val slash = thisSlash ?: otherSlash ?: Path.DIRECTORY_SEPARATOR.toSlash()
+
+  // We look at the path both have in common.
+  var lastCommonByteIndex = 0
+  while (lastCommonByteIndex < bytes.size &&
+    lastCommonByteIndex < other.bytes.size &&
+    (bytes[lastCommonByteIndex] and 0xff) == (other.bytes[lastCommonByteIndex] and 0xff)
+  ) {
+    lastCommonByteIndex++
+  }
+
+  if (lastCommonByteIndex == bytes.size && bytes.size == other.bytes.size) {
+    // `this` and `other` are the same path.
+    return ".".toPath()
+  }
+
+  if (lastCommonByteIndex == 0 && isAbsolute) {
+    throw IllegalArgumentException(
```

Comment:
Okay, so here’s a super annoying wrench. For UNC names the root is everything up to the 3rd slash, and I don’t think these can traverse to each other.

That is, I think we should throw on these paths:
```
val a = "\\\\server1\foo".toPath()
val b = "\\\\server2\bar".toPath()
```

Lemme independently verify this though.

Reply:
On the JVM:
```
\\localhost\c$\Windows
okio exists: true
file exists: true

\\127.0.0.1\c$\Windows
okio exists: true
file exists: true

\\127.0.0.1\c$\Windows\..\Windows
okio exists: true
file exists: true

\\127.0.0.1\c$\..\c$\Windows
okio exists: false
file exists: false

\\localhost\..\127.0.0.1\c$\Windows
okio exists: java.nio.file.FileSystemException: \\localhost\..\127.0.0.1\c$\Windows: The network name cannot be found.
file exists: false

\\127.0.0.1\..\localhost\c$\Windows
okio exists: java.nio.file.FileSystemException: \\127.0.0.1\..\localhost\c$\Windows: The network name cannot be found.
file exists: false

\\127.0.0.1\c$\..\..\localhost\c$\Windows
okio exists: false
file exists: false
```

Okio native:
```
\\localhost\c$\Windows
okio exists: true

\\127.0.0.1\c$\Windows
okio exists: true

\\127.0.0.1\c$\Windows\..\Windows
okio exists: true

\\127.0.0.1\c$\..\c$\Windows
okio exists: false

\\localhost\..\127.0.0.1\c$\Windows
okio exists: false

\\127.0.0.1\..\localhost\c$\Windows
okio exists: false

\\127.0.0.1\c$\..\..\localhost\c$\Windows
okio exists: false
```

Reply:
(In these I’ve manually removed support for dropping the `..` operator in `String.toPath()`.)